### PR TITLE
update: DSC status update and delete component CR if it is not set Managed

### PR DESCRIPTION
**Pull request by zdtsw**
- delete component CR when it is either set to Removed or not set in DSC
- remove old status updates"Component is disabled/enabled), only update when component CR create successful or failed (should have follow up to add check if CR status is ready or notready)
- fix a bug in datasciencpipeline :D  https://github.com/opendatahub-io/opendatahub-operator/pull/1340

## Description

split out changes not related to modelmesh from https://github.com/opendatahub-io/opendatahub-operator/pull/1338

## Merge criteria

- [ ] You have read the [contributors guide](https://github.com/opendatahub-io/opendatahub-operator/blob/incubation/CONTRIBUTING.md).
- [ ] Commit messages are meaningful - have a clear and concise summary and detailed explanation of what was changed and why.
- [ ] Pull Request contains a description of the solution, a link to the JIRA issue, and to any dependent or related Pull Request.
- [ ] Testing instructions have been added in the PR body (for PRs involving changes that are not immediately obvious).
- [ ] The developer has manually tested the changes and verified that the changes work

### CODE REVIEW

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 1]
```diff
@@ -277,37 +280,20 @@ func (r *DataScienceClusterReconciler) ReconcileComponent(
...
-	}
-
-	if err != nil {
-		r.Log.Error(err, "Failed to reconcile component: "+componentName)
-		instance = r.reportError(err, instance, fmt.Sprintf("failed to reconcile %s on DataScienceCluster", componentName))
+		r.Log.Error(err, "Failed to create component CR: "+componentName)
+		instance = r.reportError(err, instance, fmt.Sprintf("failed to create %s by DSC", componentName))
 		instance, _ = status.UpdateWithRetry(ctx, r.Client, instance, func(saved *dscv1.DataScienceCluster) {
-			status.SetComponentCondition(&saved.Status.Conditions, componentName, status.ReconcileFailed, fmt.Sprintf("Component reconciliation failed: %v", err), corev1.ConditionFalse)
+			status.SetComponentCondition(&saved.Status.Conditions, componentName, status.ReconcileFailed, fmt.Sprintf("Component creation failed: %v", err), corev1.ConditionFalse)
```
Comment (lburgazzoli) [inline controllers/datasciencecluster/datasciencecluster_controller.go:289]: nit: this method does not only creates a component, but would also update it so I think `reconciliation` is more appropriate

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 2]
```diff
@@ -78,6 +78,9 @@ type DataScienceClusterConfig struct {
 
 const (
 	finalizerName = "datasciencecluster.opendatahub.io/finalizer"
+	RHOAIAppNS    = "redhat-ods-applications"
+	ODHAppNS      = "opendatahub"
+	MonitoringNS  = "redhat-ods-monitoring"
```
Comment (lburgazzoli) [inline controllers/datasciencecluster/datasciencecluster_controller.go:83]: Those changes do not seem to be related with the PR :)

Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:83]: yeah, the lint is annoying, to get rid of the `//nolint:goconst`
but i can revert it

File: controllers/datasciencecluster/datasciencecluster_controller.go (hunk 1 above)
Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:289]: correct. let me revert it

File: controllers/datasciencecluster/datasciencecluster_controller.go (hunk 2 above)
Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go]: reverted

File: controllers/datasciencecluster/datasciencecluster_controller.go (hunk 1 above)
Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go]: updated

File: controllers/datasciencecluster/datasciencecluster_controller.go [hunk 3]
```diff
@@ -360,7 +352,7 @@ func (r *DataScienceClusterReconciler) apply(ctx context.Context, dsc *dscv1.Dat
 	}
 
 	managementStateAnn, exists := obj.GetAnnotations()[annotations.ManagementStateAnnotation]
-	if exists && managementStateAnn == string(operatorv1.Removed) {
+	if exists && (managementStateAnn == string(operatorv1.Removed) || managementStateAnn == "Unknown") {
```
Comment (lburgazzoli) [inline controllers/datasciencecluster/datasciencecluster_controller.go:355]: When is this case happening ? I wonder if we should only handle `Managed` and `Removed`, and return an error or skip otherwise ?

Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:355]: we can set it in DSC as
```
  components:
    codeflare: {}
```
which is the same
```
  components:
    codeflare:
        managementState: Removed
```

Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:355]: currently we tell user, do not set managementState in the component is the same as set it to Removed.
to make it easier i always create DSC with 
```
spec:
  components:
    modelmeshserving: 
         managementState: Managed
```
only ;D , then i found other new components get created

Comment (lburgazzoli) [reply controllers/datasciencecluster/datasciencecluster_controller.go:355]: so I guess the mapping could be done in the components' `NewCRObject` ?

Comment (zdtsw) [reply controllers/datasciencecluster/datasciencecluster_controller.go:355]: yeah, either do in all components NewCRObject(), or a central place.

thinking it out, if one day, we add Force or Unmanaged as supported case, we will need it in NewCRObject(), then DSC need logic to handle that as well.

Comment (lburgazzoli) [reply controllers/datasciencecluster/datasciencecluster_controller.go:355]: Now since components have `GetManagementState`, maybe we can move it to a central place.

File: controllers/components/dashboard/dashboard.go [hunk 4]
```diff
@@ -25,8 +25,16 @@ func (s *componentHandler) GetName() string {
...
-	return dsc.Spec.Components.Dashboard.ManagementState
+func (s *componentHandler) GetManagementState(dsc *dscv1.DataScienceCluster) string {
+	if s == nil {
+		return "Removed"
+	}
+	switch dsc.Spec.Components.Dashboard.ManagementState {
+	case operatorv1.Managed, operatorv1.Removed:
+		return string(dsc.Spec.Components.Dashboard.ManagementState)
+	default: // Force and Unmanaged case for unknown values, we do not support these yet
+		return "Unknown"
```
Comment (lburgazzoli) [inline controllers/components/dashboard/dashboard.go:36]: Wonder if this should return `operatorv1.Removed`, in other words, how do we generally handle the case where the `ManagementState` is not set ?

File: controllers/components/datasciencepipelines/datasciencepipelines.go [hunk 5]
```diff
@@ -26,11 +26,19 @@ func init() { //nolint:gochecknoinits
...
 func (s *componentHandler) GetName() string {
-	return componentsv1.DashboardComponentName
+	return componentsv1.DataSciencePipelinesComponentName
 }
 
-func (s *componentHandler) GetManagementState(dsc *dscv1.DataScienceCluster) operatorv1.ManagementState {
-	return dsc.Spec.Components.Dashboard.ManagementState
+func (s *componentHandler) GetManagementState(dsc *dscv1.DataScienceCluster) string {
+	if s == nil {
+		return "Removed"
```
Comment (lburgazzoli) [inline controllers/components/datasciencepipelines/datasciencepipelines.go:34]: I would do `return string(operatorv1.Removed)` so it is always consistent with the constant

Comment (zdtsw) [reply controllers/components/datasciencepipelines/datasciencepipelines.go:34]: i first made the one with return string(operatorv1.Removed)
then i thought to convert a const to string, is that extra step to just return a string.
but yeah, i get your point, i can get that reverted

File: pkg/componentsregistry/componentsregistry.go [hunk 6]
```diff
@@ -21,7 +20,7 @@ type ComponentHandler interface {
 	// GetName and GetManagementState sound like pretty much the same across
 	// all components, but I could not find a way to avoid it
 	GetName() string
-	GetManagementState(dsc *dscv1.DataScienceCluster) operatorv1.ManagementState
+	GetManagementState(dsc *dscv1.DataScienceCluster) string
```
[the remark this answers is not in the document]
Comment (zdtsw) [reply pkg/componentsregistry/componentsregistry.go:23]: yeah, lets revert back to operatorv1.ManagementState

File: controllers/components/dashboard/dashboard.go (hunk 4 above)
Comment (zdtsw) [reply controllers/components/dashboard/dashboard.go:36]: i think it falls into the `s == nil `case ?

Comment (zdtsw) [reply controllers/components/dashboard/dashboard.go:36]: Now, i think i understand your comments better:
i can think of two approaches:
1. remove this default for now: since enum is only for Removed and Managed. till we add support for Force and Unmanaged, we do not need to worry about "default"
2. change the whole funciton to
```
func (s *componentHandler) GetManagementState(dsc *dscv1.DataScienceCluster) operatorv1.ManagementState {
	if dsc.Spec.Components.Dashboard.ManagementState == operatorv1.Managed {
		return operatorv1.Managed
	}
	return operatorv1.Removed
}
```

Comment (lburgazzoli) [reply controllers/components/dashboard/dashboard.go:36]: yep something like option 2

Comment (zdtsw) [reply controllers/components/dashboard/dashboard.go]: updated

**Review by lburgazzoli — APPROVED** [review]
